sdks/python/apache_beam/transforms/trigger_test.py

```diff
@@ -659,11 +675,132 @@ def fire_timers():
     self.assertEqual([], output, msg='Unexpected output: %s' % output)
 
 
+class TestStreamTranscriptTest(TranscriptTest):
+  """A suite of TestStream-based tests based on trigger transcript entries.
+  """
+
+  def _execute(
+      self, window_fn, trigger_fn, accumulation_mode, timestamp_combiner,
+      transcript, spec):
+
+    runner_name = TestPipeline().runner.__class__.__name__
+    if runner_name in spec.get('broken_on', ()):
+      self.skipTest('Known to be broken on %s' % runner_name)
+
+    test_stream = TestStream()
+    for action, params in transcript:
+      if action == 'expect':
+        test_stream.add_elements([('expect', params)])
+      else:
+        test_stream.add_elements([('expect', [])])
+        if action == 'input':
+          test_stream.add_elements([('input', e) for e in params])
+        elif action == 'watermark':
+          test_stream.advance_watermark_to(params)
+        elif action == 'clock':
+          test_stream.advance_processing_time(params)
+        elif action == 'state':
+          pass  # Requires inspection of implementation details.
+        else:
+          raise ValueError('Unexpected action: %s' % action)
+    test_stream.add_elements([('expect', [])])
```

Comment:
What does this line do? Always add an empty expectation at the end?

Reply:
Exactly.

Reply:
I should've rephrased the question: What do we need this empty expectation for? Also why do we need an empty expectation for all the other actions at line 696?

Reply:
We want to make sure no elements are produced where they're not expected. E.g. if I had a transcript that looked like

```
window_fn: FixedWindows(10)
transcript:
  - input [1, 2, 3]
  - watermark: 5
  - watermark: 6
```

I would want to assert that nothing came out when the watermark was advanced to 5 or 6